src/bp/core/services/action/action-service.ts

```diff
@@ -38,6 +57,23 @@ export default class ActionService {
     this._invalidateDebounce = _.debounce(this._invalidateRequire, DEBOUNCE_DELAY, { leading: true, trailing: false })
   }
 
+  forBot(botId: string): ScopedActionService {
+    if (this._scopedActions.has(botId)) {
+      return this._scopedActions.get(botId)!
+    }
+
+    const scopedActionService = new ScopedActionService(
+      this.ghost,
+      this.logger,
+      botId,
+      this.cache,
+      this.tasksRepository,
+      this.workspaceService
+    )
+    this._scopedActions.set(botId, scopedActionService)
```

Comment:
This will make `forBot()` async, and I don't think we want that.

Reply:
What if we make it async ? That service is not exposed in the SDK and we use it in at most 3 places. I think this would make sense. 

```js
async forBot(botId: string): Promise<ScopedActionService> {
    if (this._scopedActions.has(botId)) {
      return this._scopedActions.get(botId)!
    }

    if (!(await this.botService.botExists(botId))) {
      throw new NotFoundError(`This bot doesn't exist`)
    }

    const workspaceId = await this.workspaceService.getBotWorkspaceId(botId)

    const scopedGhost = new ScopedActionService(this.ghost, this.logger, botId, this.cache, workspaceId)
    this._scopedActions.set(botId, scopedGhost)
    return scopedGhost
  }
```

Then in the router you can consume it this way:

```js
this.app.post('/action/run', _validateRunRequest, asyncMiddleware(async (req, res) => {
    const { incomingEvent, actionArgs, actionName, botId, token } = req.body

    await this.actionService.forBot(botId).then(a => a.runLocalAction({ actionName, actionArgs, incomingEvent, token, runType: 'http' }))

    res.send({ incomingEvent })
  })
```

You don't even need to wrap it in a try catch when you wrap it in asyncMiddleware, because it handles seamlessly HTTP errors (our custom NotFoundError)

Reply:
Fixed in a4546d9